src/lem/circuit.rs

```diff
@@ -0,0 +1,771 @@
+//! `LEMOP::slots_info`;
+//!
+//! 2. Interpret the LEM and collect the data that was generated on each visited
+//! slot, along with all bindings from `MetaPtr`s to `Ptr`s. This piece of
+//! information will live on a `Frame` structure;
+//!
+//! 3. Finally build the circuit with `SlotsInfo` and `Frame` at hand. This step
+//! is explained in more details in the `LEMOP::synthesize` function.
+//!
+//! The 3 steps above will be further referred to as *STEP 1*, *STEP 2* and
+//! *STEP 3* respectively. STEP 1 should be performed once per slot. Then STEP 2
+//! will need as many iterations as it takes to evaluate the Lurk expression and
+//! so will STEP 3.
+
+use std::collections::HashMap;
+
+use anyhow::{anyhow, bail, Context, Result};
+use bellperson::{
+    gadgets::{boolean::Boolean, num::AllocatedNum},
+    ConstraintSystem,
+};
+use indexmap::{IndexMap, IndexSet};
+
+use crate::circuit::gadgets::{
+    constraints::{
+        alloc_equal_const, and, enforce_selector_with_premise, implies_equal, implies_equal_zero,
+    },
+    data::hash_poseidon,
+    pointer::AllocatedPtr,
+};
+
+use crate::field::{FWrap, LurkField};
+
+use super::{
+    interpreter::{Frame, Slot, SlotType},
+    path::Path,
+    pointers::ZPtr,
+    store::Store,
+    MetaPtr, LEM, LEMOP,
+};
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SlotsCounter {
+    pub hash2: usize,
+    pub hash3: usize,
+    pub hash4: usize,
+}
+
+impl SlotsCounter {
+    /// This interface is mostly for testing
+    #[inline]
+    #[allow(dead_code)]
+    pub(crate) fn new(num_slots: (usize, usize, usize)) -> Self {
+        Self {
+            hash2: num_slots.0,
+            hash3: num_slots.1,
+            hash4: num_slots.2,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn next_hash2(&self) -> Self {
+        Self {
+            hash2: self.hash2 + 1,
+            hash3: self.hash3,
+            hash4: self.hash4,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn next_hash3(&self) -> Self {
+        Self {
+            hash2: self.hash2,
+            hash3: self.hash3 + 1,
+            hash4: self.hash4,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn next_hash4(&self) -> Self {
+        Self {
+            hash2: self.hash2,
+            hash3: self.hash3,
+            hash4: self.hash4 + 1,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn max(&self, other: Self) -> Self {
+        use std::cmp::max;
+        Self {
+            hash2: max(self.hash2, other.hash2),
+            hash3: max(self.hash3, other.hash3),
+            hash4: max(self.hash4, other.hash4),
+        }
+    }
+}
+
+/// Contains a `slots_map` that maps `LEMOP`s to their slots. This map is not
+/// expected to be injective, as two or more `LEMOP`s can be mapped to the same
+/// slot. The `slots` attribute is an set of all slots present on `slots_map`.
+pub struct SlotsInfo {
+    pub slots_map: IndexMap<LEMOP, Slot>,
```

Comment:
The existence of this map points to unnecessary complexity. The slot mechanism shouldn't require any information about 'who' (i.e. the LEMOP) is using the slot. All that matters is *what* is being done, and the order in which it happens — which is fixed by evaluation order.

Reply:
I had implemented a solution closer to that before but I realized that it created information/assumptions that forced some entanglement between interpretation and circuit generation. That entanglement is undone by capturing this information in a map, which frees us from thinking about ordering or interactions between traversal orderings in different places.